File: app/src/main/java/com/mxt/anitrend/adapter/recycler/shared/GroupMediaViewHolder.java
package com.mxt.anitrend.adapter.recycler.shared;

import android.view.View;

import com.annimon.stream.IntPair;
import com.bumptech.glide.Glide;
import com.mxt.anitrend.R;
import com.mxt.anitrend.base.custom.recycler.RecyclerViewHolder;
import com.mxt.anitrend.base.interfaces.event.ItemClickListener;
import com.mxt.anitrend.databinding.AdapterMediaHeaderBinding;
import com.mxt.anitrend.model.entity.base.MediaBase;
import com.mxt.anitrend.model.entity.group.RecyclerItem;

import butterknife.OnClick;
import butterknife.OnLongClick;

/**
 * Created by max on 2018/03/26.
 * Group header for media items
 */

public class GroupMediaViewHolder extends RecyclerViewHolder<RecyclerItem> {

    private AdapterMediaHeaderBinding binding;
    private ItemClickListener<RecyclerItem> clickListener;

    /**
     * Default constructor which includes binding with butter knife
     *
     * @param binding
     */
    public GroupMediaViewHolder(AdapterMediaHeaderBinding binding, ItemClickListener<RecyclerItem> clickListener) {
        super(binding.getRoot());
        this.binding = binding;
        this.clickListener = clickListener;
    }

    /**
     * Load image, text, buttons, etc. in this method from the given parameter
     * <br/>
     *
     * @param recyclerItem Is the model at the current adapter position
     */
    @Override
    public void onBindViewHolder(RecyclerItem recyclerItem) {
        MediaBase model = (MediaBase) recyclerItem;
        binding.setModel(model);
        binding.seriesTitle.setTitle(model);
        binding.executePendingBindings();
    }

    /**
     * If any image views are used within the view holder, clear any pending async img requests
     * by using Glide.clear(ImageView) or Glide.with(context).clear(view) if using Glide v4.0
     * <br/>
     *
     * @see Glide
     */
    @Override
    public void onViewRecycled() {
        Glide.with(getContext()).clear(binding.seriesImage);
        binding.unbind();
    }

    @Override @OnClick(R.id.container)
    public void onClick(View v) {
        IntPair<Boolean> pair = isValidIndexPair();
        if(binding != null && binding.getModel() != null && isClickable(binding.getModel()) && pair.getSecond())
            clickListener.onItemClick(v, new IntPair<>(pair.getFirst(), binding.getModel()));
    }

    @Override @OnLongClick(R.id.container)
    public boolean onLongClick(View v) {
        IntPair<Boolean> pair = isValidIndexPair();
        if(binding != null && binding.getModel() != null && isLongClickable(binding.getModel()) && pair.getSecond()) {
            clickListener.onItemLongClick(v, new IntPair<>(pair.getFirst(), binding.getModel()));
            return true;
        }
        return false;
    }
}

File: app/src/main/java/com/mxt/anitrend/adapter/recycler/shared/GroupMediaViewHolder.kt
package com.mxt.anitrend.adapter.recycler.shared

import android.view.View
import com.bumptech.glide.Glide
import com.mxt.anitrend.R
import com.mxt.anitrend.base.custom.recycler.RecyclerViewHolder
import com.mxt.anitrend.base.custom.view.image.AspectImageView
import com.mxt.anitrend.base.custom.view.widget.SeriesStatusWidget
import com.mxt.anitrend.base.interfaces.event.ItemClickListener
import com.mxt.anitrend.databinding.AdapterMediaHeaderBinding
import com.mxt.anitrend.model.entity.base.MediaBase
import com.mxt.anitrend.model.entity.group.RecyclerItem
import com.mxt.anitrend.util.CompatUtil

/**
 * Created by max on 2018/03/26.
 * Group header for media items
 */
class GroupMediaViewHolder(
    private val binding: AdapterMediaHeaderBinding,
    private val clickListener: ItemClickListener<RecyclerItem>,
) : RecyclerViewHolder<RecyclerItem>(binding.root) {
    private var boundModel: MediaBase? = null

    init {
        bindClickListeners(R.id.container)
        bindLongClickListeners(R.id.container)
    }

    override fun onBindViewHolder(model: RecyclerItem) {
        val media = model as? MediaBase ?: return
        boundModel = media
        AspectImageView.setImage(binding.seriesImage, media.coverImage)
        SeriesStatusWidget.setStatus(binding.seriesStatus, media)
        binding.seriesTitle.setTitle(media)
        val subGroupTitle = media.subGroupTitle
        binding.seriesSubgroupTitle.text =
            if (subGroupTitle.isNullOrBlank()) "" else CompatUtil.capitalizeWords(subGroupTitle)
        val format = media.format
        binding.seriesFormat.text =
            if (format.isNullOrBlank()) "" else CompatUtil.capitalizeWords(format)
    }

    override fun onViewRecycled() {
        Glide.with(getContext()).clear(binding.seriesImage)
        boundModel = null
    }

    override fun onClick(v: View) {
        val pair = isValidIndexPair()
        val model = boundModel
        if (model != null && isClickable(model) && pair.value) {
            clickListener.onItemClick(v, IndexedValue(pair.index, model))
        }
    }

    override fun onLongClick(v: View): Boolean {
        val pair = isValidIndexPair()
        val model = boundModel
        if (model != null && isLongClickable(model) && pair.value) {
            clickListener.onItemLongClick(v, IndexedValue(pair.index, model))
            return true
        }
        return false
    }
}
